book/src/solana-scripts.md

```diff
@@ -0,0 +1,253 @@
+
+Since scripts are clients, they cannot:
+* Directly mutate state, or directly move funds between accounts.
+
+Scripts rely on programs to move funds or mutate state, and act purely as
+clients.
+
+## Secure Script Execution
+
+During the script execution, calls to `process_instruction` yield, and the next
+instruction to be processed is invoked.  Scripts execute dynamically, and the
+execution path based on the current on-chain state.  The state may not match the
+same state that clients observed when signing the transaction.  Since script
+execute multiple programs, any change to any of the programs in the execution
+path could potentially change the outcome of the script.
+
+Several solutions exist to this problem:
+
+* clients sign a transaction with a specific blockhash, if any of the state that
+the script depends on have changed signed the signature the transaction fails.
+
+This solution is problematic because the stable blockhash is the root, but the
+transaction will be encoded at the head.  So the maximum throughput per script
+is going to be limited by the speed at which each transaction finalizes with a
+high confidence, which could be at least 10 blocks.
+
+* clients sign the expected merkle root of the changes to state after the
+transaction.  If the changes fail to match, the transaction fails.
+
+This solution has better performance characteristics, but users must serialize
+all the transactions to the script, because changes cannot be applied out of
+order.
+
+This design proposes another approach.  Users know ahead of time which
+instructions the script will generate, and if signatures are required for the
+instructions.
+
+The transaction invoked by the client must declare all the accounts that the
+script will need up front and provide all the necessary client signatures, as
+well as encode the instruction vector that the script will generate.
+
+The latter provides clear authorization for the script to take actions on behalf
+of the client, since each instruction specifies the clients keys and explicit
+signatures for each explicit instruction.  Users do not need to guess which
+instructions the script will execute, and authorize each one explicitly.  If the
+on-chain state has changed such that the script will take a different branch and
+execute a different instruction, then the script fails.  This allows the
+programs to implement out of order composition and for users to take advantage
+of those performance benefits.
+
+Once the instruction is invoked, the script is resumed from the last point of
+execution.
+
+## Methods
+
+A TTP needs the ability to create pubkeys and generate signatures for those
+keys.  At script creation time the loader program may authorize specific
+pubkeys that the script and only the script can sign with.  To ensure that these
+keys cannot be signed by the client, the addresses are derived from a sha256 of
+the script pubkey and the key sequence number.
```

Comment:
What is a key sequence number?

Reply:
@jackcmay yea, its a key index.  think of it as a local wallet with `N` keys.